lib/Sema/CSGen.cpp

```diff
@@ -1783,7 +1783,53 @@ namespace {
                                         expr->getElementName(i), flags));
       }
 
-      return TupleType::get(elements, CS.getASTContext());
+      auto &ctx = CS.getASTContext();
+      auto locator = CS.getConstraintLocator(expr);
+      auto contextualTy = CS.getContextualType(expr);
+      auto tupleTy = TupleType::get(elements, ctx);
+
+      // Somewhat Fast path: if contextual type is a tuple, return the tuple.
+      if (contextualTy && contextualTy->is<TupleType>()) {
```

Comment:
I would express this as a conversion, maybe even tack it onto `LiteralConformsTo`.  Going after the contextual type in CSGen is a bit too early in constraint solving.

Reply:
Does `LiteralConformsTo` make sense here? Because we aren’t introducing a new `ExpressibleBy*Literal`, so I’m not sure what the protocolTy would be.

Reply:
Sorry, that was unclear.  Let me rephrase: This constraint makes the most sense as a conversion constraint in the style of `LiteralConformsTo` for an infinite family of ad-hoc `ExpressibleByTupleOfXXXLiteral` protocols.  You could emit such a constraint and basically nuke everything down below here.  Then your corresponding `CSSimplify` rule breaks this down into an argument match against constructors.